honeybadgermpc/polynomial.py

```diff
@@ -277,6 +277,124 @@ def fft(poly, omega, n, seed=None):
     return fft_helper(padded_coeffs, omega, poly.field)
 
 
+def fnt_decode_step1(poly, zs, omega2, n):
+    """
+    This needs to be run once for decoding a batch of secret shares
+    It depends only on the x values (the points the polynomial is
+    evaluated at, i.e. the IDs of the parties contributing shares) so
+    it can be reused for multiple batches.
+    Complexity: O(n^2)
+
+    args:
+        zs is a subset of [0,n)
+        omega2 is a (2*n)th root of unity
+
+    returns:
+        A(X) evaluated at 1...omega2**(2n-1)
+        Ai(xi) for each xi = omega**(zi)
+
+    where:
+        omega = omega2**2
+        where A(X) = prod( X - xj ) for each xj
+        Ai(xi) = prod( xi - xj ) for j != i
+    """
+    k = len(zs)
+    omega = omega2**2
+    xs = [omega**z for z in zs]
+
+    # Compute A(X)
+    a_ = poly([1])
+    for i in range(k):
+        print(i, poly([-xs[i], 1]))
+        a_ *= poly([-xs[i], 1])
+    as_ = [a_(omega2**i) for i in range(2*n)]
+
+    # Compute all Ai(Xi)
+    ais_ = []
+    for i in range(k):
+        ai = a_.field(1)
+        for j in range(k):
+            if i != j:
+                ai *= xs[i] - xs[j]
+        ais_.append(ai)
+    return as_, ais_
+
+
+def fnt_decode_step2(poly, zs, ys, as_, ais_, omega2, n):
+    """
+    Returns a polynomial P such that P(omega**zi) = yi
+
+    Complexity: O(n log n)
+
+    args:
+        zs is a subset of [0,n)
+        As, Ais = fnt_decode_step1(zs, omega2, n)
+        omega2 is a (2*n)th root of unity
+
+    returns:
+        P  Poly
+    """
+    k = len(ys)
+    assert len(ys) == len(ais_)
+    assert len(as_) == 2 * n
+    omega = omega2 ** 2
+
+    # Compute N'(x)
+    nis = [ys[i] / ais_[i] for i in range(k)]
+    ncoeffs = [0 for _ in range(n)]
+    for i in range(k):
+        ncoeffs[zs[i]] = nis[i]
+    n_ = poly(ncoeffs)
+
+    # Compute P/A(X)
+    # PoverA = -Poly([N(omega**(n-j-1)) for j in range(n)])
+    nevals = n_.evaluate_fft(omega, n)
+    power_a = -poly(nevals[::-1])
+    pas = power_a.evaluate_fft(omega2, 2*n)
+
+    # Recover P(X)
+    ps = [p * a for (p, a) in zip(pas, as_)]
+    prec = poly.interpolate_fft(ps, omega2)
+    prec.coeffs = prec.coeffs[:k]
+    return prec
+
+
+class EvalPoint(object):
+    """Helper to generate evaluation points for polynomials between n parties
+
+    If FFT is being used:
+    omega is a root of unity s.t. order(omega) = (smallest power of 2 >= n)
+    i'th point (zero-indexed) = omega^(i + 1)
+
+    Without FFT:
+    i'th point (zero-indexed) = i + 1
+    """
+    def __init__(self, field, n, use_fft=False):
+        self.use_fft = use_fft
+        self.field = field
+
+        # Need an additional point where we evaluate polynomial to get secret
+        order = n
+        if use_fft:
+            self.order = order if (order & (order - 1) == 0) else 2 ** order.bit_length()
+
+            self.omega2 = get_omega(field, 2 * self.order, seed=0)
```

Comment:
any reason why we want to fix the seed?

Reply:
All parties must arrive at the same omega for the algorithm to work. Fixing the seed was the easiest way to do that that I could think of.

Reply:
I'll document this in the code.